### PR TITLE
Block Checkout: Fix default shipping selection on toggle pickup

### PR DESCRIPTION
**Pull request by mikejolley**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Fixes a regression from https://github.com/woocommerce/woocommerce/pull/48370 noticed by @nielslange whereby toggling between shipping and pickup would not reselect a shipping method automatically. 

The reason for this is unlike with pickup locations, the component forces the default state to be the first shipping rate in the array. Because of this, the event is not pushed up to the data store.

By defaulting to a blank default shipping rate, the mounting code can select the shipping rate instead and push the change upward. This matches up pickup locations do it.

Additionally, there was an issue where this notice would flash up when changing between shipping and pickup:

Screenshot 2024-07-19 at 14 00 27

This can be fixed by removing the notice during calculations.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Setup some shipping methods and some local pickup locations under WooCommerce > Settings > Shipping. Additionally, create a zone for the US with no shipping methods. We'll use this later.
2. Go to checkout.
3. Using a non-US address,Toggle between local pickup and shipping. Ensure when you toggle, the first shipping rate or pickup location is automatically selected and totals are updated as expected. Also ensure that the notice "Totals will be recalculated when a valid shipping method is selected." does not become visible.
4. Switch to shipping and enter a US address (ensure you select a state also in the form as there is an unrelated issue with the select boxes preventing default state from being selected).
5. Check that the "Totals will be recalculated when a valid shipping method is selected." method appears and there are no shipping methods available.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/shipping-rates-control-package/package-rates.tsx [hunk 1]
```diff
@@ -38,15 +38,9 @@ const PackageRates = ( {
...
-		if ( selectedRateId ) {
-			return selectedRateId;
-		}
-		// Default to first rate if no rate is selected.
-		if ( rates.length > 0 ) {
-			return rates[ 0 ].rate_id;
-		}
-	} );
+	const [ selectedOption, setSelectedOption ] = useState(
+		selectedRateId ?? ''
```
Comment (mikejolley) [inline plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/shipping-rates-control-package/package-rates.tsx:42]: Empty initial state causes the mounting code on line 65 to select the initial rate. cc @senadir

File: plugins/woocommerce-blocks/assets/js/base/context/hooks/use-show-shipping-total-warning.ts [hunk 2]
```diff
@@ -18,64 +16,36 @@ export const useShowShippingTotalWarning = () => {
...
 			// Early return because shipping rates were not yet loaded from the cart data store, or the user is changing
 			// rate, no need to alter the notice until we know what the actual rate is.
+			if ( isShowingNotice ) {
+				// Removes the notice in case it was already shown.
+				removeNotice( errorNoticeId, context );
+			}
 			return;
 		}
 
-		const selectedRates = cartData?.shippingRates?.reduce(
```
Comment (mikejolley) [inline plugins/woocommerce-blocks/assets/js/base/context/hooks/use-show-shipping-total-warning.ts:46]: All of this code is replaced by the newer `hasSelectedLocalPickup`

File: plugins/woocommerce-blocks/assets/js/base/context/hooks/use-show-shipping-total-warning.ts [hunk 3]
```diff
@@ -18,64 +16,36 @@ export const useShowShippingTotalWarning = () => {
...
 	useEffect( () => {
+		const isShowingNotice = shippingNotices.length > 0;
+		const hasMismatch = ! prefersCollection && hasSelectedLocalPickup;
+
 		if ( ! hasRates || isRateBeingSelected ) {
 			// Early return because shipping rates were not yet loaded from the cart data store, or the user is changing
 			// rate, no need to alter the notice until we know what the actual rate is.
+			if ( isShowingNotice ) {
+				// Removes the notice in case it was already shown.
+				removeNotice( errorNoticeId, context );
```
Comment (mikejolley) [inline plugins/woocommerce-blocks/assets/js/base/context/hooks/use-show-shipping-total-warning.ts:42]: This remove prevents the notice from being stuck when toggling.

**Review by nielslange — APPROVED** [review]
Thanks for fixing this bug, @mikejolley. The PR works as expected and the code looks great. That said, let's wait for @senadir to double-check [your comment](https://github.com/woocommerce/woocommerce/pull/49718/files/5bd654f059ce03b0f34b86beca3e2107f72f4735#r1684352649) before shipping this fix.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/shipping-rates-control-package/package-rates.tsx (hunk 1 above)
Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/shipping-rates-control-package/package-rates.tsx:42]: The logic above would select the selected one or the first one, why are we defaulting to `""`? this means having no selection. Do you mean the hook below would handle that?

Comment (mikejolley) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/shipping-rates-control-package/package-rates.tsx:42]: @senadir Look at the file as a whole and you'll see we have some on mount logic which selects the first method and ensures the on change event is fired. If we default to a rate here, no on change event is fired. So the data store would not be in sync with the component. Note local pickup uses `''` too for the same reason.

**Pull request merged**